File: app/src/main/java/com/app/sid/funwithflags/view/countrylist/CountryListMvp.java
package com.app.sid.funwithflags.view.countrylist;

import com.app.sid.funwithflags.datasets.remote.Country;

import java.util.List;

import rx.Observable;

public interface CountryListMvp {

    interface View {
        void showProgress(boolean isVisible);

        void showInternetError(boolean isVisible);

        boolean isInternetAvailable();

        void updateCountries(List<Country> countries);

        void onQueryTextSubmit(String query);

        void onQueryTextChange(String query);
    }

    interface Interactor {
        Observable<List<Country>> getCountryList();
    }

    interface RemoteDataSource {
        Observable<List<Country>> getCountryList();
    }

    interface LocalDataSource {
        boolean isLocalDataPresent();

        Observable<List<Country>> getCountryList();

        void saveCountry(Country country);
    }
}

File: app/src/main/java/com/app/sid/funwithflags/view/countrylist/CountryListMvp.kt
package com.app.sid.funwithflags.view.countrylist

import com.app.sid.funwithflags.datasets.remote.Country
import rx.Observable

interface CountryListMvp {
    interface View {
        fun showProgress(isVisible: Boolean)
        fun showInternetError(isVisible: Boolean)
        fun isInternetAvailable(): Boolean
        fun updateCountries(countries: List<Country>)
        fun onQueryTextSubmit(query: String)
        fun onQueryTextChange(query: String)
    }

    interface Interactor {
        fun getCountryList(): Observable<List<Country>>?
    }

    interface RemoteDataSource {
        fun getCountryList(): Observable<List<Country>>?
    }

    interface LocalDataSource {
        fun isLocalDataPresent(): Boolean
        fun getCountryList(): Observable<List<Country>>?
        fun saveCountry(country: Country?)
    }
}